### ISSUE TITLE
Feature Request: Model output as a Dictionary

### ISSUE THREAD

**Issue by Jacyle** [tgsmith61591/pmdarima#54]
#### Description
By outputting the model summary statistics, it will allow for passing descriptive statistics like p-value, R-squared, coefficients, the model specifications (ex. 'order = (1,1,1)'), etc... as parameters to other parts of your code. This can be useful for setting logic and thresholds for model requirements (ex. if r_sq >= 0.80).

**Comment by tgsmith61591** [comment]
Problem setup:

```python

from pmdarima.datasets import load_lynx
from pmdarima.arima import auto_arima
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt
import numpy as np

# Load the data and split it into separate pieces
data = load_lynx()
train, test = data[:90], data[90:]

# Fit a simple auto_arima model
modl = auto_arima(train, start_p=1, start_q=1, start_P=1, start_Q=1,
                  max_p=5, max_q=5, max_P=5, max_Q=5, seasonal=True,
                  stepwise=True, suppress_warnings=True, D=10, max_D=10,
                  error_action='ignore')
```

Are you looking for something like this?

```python
>>> {'pvalues': modl.pvalues(), 'resid': modl.resid(), 'order': modl.order,
     'seasonal_order': modl.seasonal_order, 'oob': modl.oob(), 'aic': modl.aic(),
     'aicc': modl.aicc(), 'bic': modl.bic(), 'bse': modl.bse(), 'params': modl.params()}
{'pvalues': array([2.04752445e-03, 1.43710465e-61, 1.29504002e-10, 5.22119887e-15]), 'resid': array([-1244.3973072 ,  -302.89697033,  -317.63342593,  -304.57267897,
         131.69413491,   956.15566697,   880.37459722,  2445.86460353,
        -192.84268876,  -177.1932523 ,  -101.67727903,   384.05487582,
        -304.52047818,  -570.72748088,  -497.48574217,  1286.86848903,
        -400.22840217,  1017.55518758, -1157.37024626,  -295.26213543,
         104.79931827,  -574.9867485 ,  -588.49652697,  -535.37707505,
        -355.71298419,  -164.06179682,   574.51900799,    15.45522718,
       -1358.43416826,   120.42735893,  -147.94038284,  -685.64124874,
        -365.18947057,  -243.79704985,   317.79437422,   585.59553667,
          34.70605783,  -216.21587989,  -692.53375089,   116.87379358,
        -385.52193301,  -540.95554558,  -283.16913167,   438.72324376,
        1078.63542578,  3198.50449405, -2167.76083646,  -783.80525821,
        1384.85947061,   -95.84379882,  -728.85293118,   -35.68476597,
         211.33538732,  -379.91950618,   599.42290213,  -839.30599392,
        -201.97018962,  -393.28468589,  -376.16010796,  -516.52280993,
        -369.25037143,  -362.25159504,   783.17714317,   207.96692746,
        1744.27617969, -1573.37293342,  -479.20751405,   473.18948601,
        -503.20223823,  -648.62384466,  -671.12469446,  -547.51554005,
        -501.37768686,   274.76714385,  2073.1897026 , -1063.19580729,
       -1664.39957997,   882.73400004,  -304.17429193,  -422.60267409,
        -292.34984241,   -27.76090888,  1724.60937822,  3095.90133612,
        -325.78549678,   110.95150845,   645.21273504,  -135.91225092,
         417.12710097,  -118.27553718]), 'order': (2, 0, 0), 'seasonal_order': (0, 0, 0, 1), 'oob': nan, 'aic': 1487.8850037609368, 'aicc': 1488.5992894752226, 'bic': 1497.8842424422578, 'bse': array([2.26237893e+02, 6.97744631e-02, 9.58556537e-02, 1.03225425e+05]), 'params': array([ 6.97548186e+02,  1.15522102e+00, -6.16136459e-01,  8.07374077e+05])}
```

**Comment by Jacyle** [comment]
Ya, that'd be perfect. People could do some cool stuff being able to pass the descriptive stats as parameters

**Comment by tgsmith61591** [comment]
Would you be interested in taking this on as a PR? Could be an instance function on the `ARIMA` class and would resemble something like the following:

```python
@if_has_delegate("arima_res_")
def to_dict(self):
    # Your code here...
```

**Comment by Jacyle** [comment]
I'd be willing to. But I'll need to do some reading.

You're a much more advanced programmer than I am. I'm an amateur at best. You can check out my project I started in ~February and published to Github, [binance-technical-algorithm](https://github.com/Jacyle/binance-technical-algorithm), which will give you an idea of my level of programming skills.

If you could provide me with some guidance; some things to read and maybe an overview of how you'd code it, It'd be a great learning experience for me :-)

**Comment by tgsmith61591** [comment]
No problem. There's no better way to learn than to *do*.

As far as guidance goes, I think the method signature should resemble:

```python
@if_has_delegate("arima_res_")
def to_dict(self):
    # Your code here...
```

The method will go inside the `ARIMA` class in `pmdarima/arima/arima.py`. Make sure you include a docstring that explains what the method does and returns. If you need help with the docstring, take a look at some of the other ones in the module. They usually resemble:

```python
def func():
    """One line summary

    This function literally does nothing. It returns the number one.

    Returns
    -------
    res : int
        The number one.
    """
    return 1
```

### Testing

You'll need to create a unit test to test your function. Here are several test cases I think would be valuable:

* [ ] The method returns a `dict` type (`assert isinstance(result, dict)`)
* [ ] The method fails with an `AttributeError` when an `ARIMA` has not yet been fit. You can use the [`pytest.raises`](https://docs.pytest.org/en/latest/assert.html) syntax to assert this.
* [ ] The keys that you expect are present in the dictionary

To create a unit test, go into the `pmdarima/arima/tests/test_arima.py` file and add a new function (or more) with the `test_` prefix. If the function starts with `test_`, the testing framework will find it:

```python
def test_something():
    # Will be tested

def something():
    # Will not be run by unit tests
```

See the [testing section](https://www.alkaline-ml.com/pmdarima/contributing.html#pull-request-checklist) of the pull request checklist to get an idea of how to run the tests and build from source locally.


### Making commits

When you make commits with your change, it would be helpful if your commit message included a reference to this issue (issue 54). Git is smart about this. To reference the issue in a commit, simply include `#54` in your commit message, something like:

```bash
$ git commit -m 'Add to_dict for #54'
```

### Acceptance criteria

When you push, the CI jobs will run automatically. If they all pass and the above test cases look good, we'll review and merge it in. You'll have to fork the project to push commits, and then when it's ready you'll want to to create a branch off of master to do your work:

```bash
$ git checkout -b feature_for_issue_54
```

When you submit your PR upstream, please request to merge into the `develop` branch, as we use `master` for releases only. [This guide](https://www.alkaline-ml.com/pmdarima/contributing.html#contributing-to-pmdarima) should be helpful :)

After you've created a PR, it may fail or you may decide to add more to it. All you have to do is continue making pushes to the branch you're requesting to merge and the PR will update. Might be good to read [this](https://help.github.com/articles/about-pull-requests/) (follow the **Further Reading** section at the bottom as well).

We'll follow the scikit-learn convention: when your PR is ready to go, edit the title to from `Whatever your title was` to `[MRG] Whatever your title was`, and that will signal to us that you're ready for a code review. After one of us reviews, we will change it to `[MRG + 1] Whatever your title was`. When the second reviewer takes a look and it all looks good, we'll change it to `[MRG + 2] Whatever your title was` and will merge in and this whole process will be complete.

I know this is a lot of information, but it's the best way to learn! And it's always a plus for us when we get people excited about contributing to the project.

**Comment by aaronreidsmith** [comment]
I can take this one on!

**Comment by aaronreidsmith** [comment]
This issue has been implemented by PR #61. Closing this issue
